cmd/operator-sdk/add/controller.go

```diff
@@ -57,6 +62,8 @@ Example:
 	if err := controllerCmd.MarkFlagRequired("kind"); err != nil {
 		log.Fatalf("Failed to mark `kind` flag for `add controller` subcommand as required")
 	}
+	controllerCmd.Flags().StringVar(&k8sImportPath, "k8s-import-path", "k8s.io/api", "Kubernetes resource import path. Only valid if --k8s-api is set")
```

Comment:
But if you disagree with the above, we should put a warning that the flag `k8s-import-path` the user sets will be ignored when the `k8s-api` is not set to true.

Reply:
I prefer the second method. @joelanford wdyt?

Reply:
I think there are three scenarios to cover:
1. API is provided by this project (what we currently expect)
2. API is provided by kubernetes from `k8s.io/api`
3. API is provided by another non-k8s project.

Based on that, I think we should have either one flag that supports both 2 and 3, or mutually exclusive flags.

1. In the single flag scenario, I think we could map the above scenarios to the following values with a hybrid boolean/string flag:
    1. Unset / `--use-k8s-api-import=false`
    2. `--use-k8s-api-import` / `--use-k8s-api-import=true` / `--use-k8s-api-import=""`
    3. `--use-k8s-api-import="mygroupv1=github.com/myproject/pkg/apis/mygroup/v1"`

2. In the two flag scenario, we could have:
    1. No import flags
    2. `--k8s-api=true`
    3. `--custom-api-import="mygroupv1=github.com/myproject/pkg/apis/mygroup/v1"`

In this case, if both `--k8s-api` and `--custom-api-import` are set, we would error out.

Reply:
On second thought I like @LiliC's suggestion/`1.iii` best, except it should be something more like:

```console
$ operator-sdk add api --kind Deployment --api-version core/v1 --custom-api-import-path
$ operator-sdk add api --kind CustomKind --api-version some.group.com/v1alpha1 --custom-api-import-path github.com/myproject/pkg/apis
```

since we're already passing a part of the path in `--api-version`.

This way we can use the same flag for all external API's, k8s or not. We can use the hybrid strategy with a default `k8s.io/api`, or require the user pass an import path.

Reply:
I also like that we call the flag something more generic, like `--custom-api-import-path` But we can still mention an example in the docs for the k8s api use case. 👍

Reply:
My one hesitation is that this would mean we would be opinionated about the location of types packages in external projects. Using your `CustomKind` example:

We would import:
```
github.com/myproject/pkg/apis/some/v1alpha1
```

But what if the external types were actually in:

```go
github.com/myproject/pkg/apis
// or
github.com/myproject/pkg/apis/some.group.com/v1alpha1
// or
github.com/myproject/pkg/apis/some/v1alpha1/customkind
```

Part of my concern is also that the flag name is `--custom-api-import-path` but we're treating it as a prefix. I think the most intuitive thing for a user would be to just pass the full path.

Also, do we need to enable the user to set the import alias themselves? There are two scenarios that it may be necessary:
1. It conflicts with another alias
    - `--api-version=core.example.com/v1` => `corev1`
2. The import path may not be conducive to the way we would derive the alias.
    - `--custom-api-import-path=github.com/myproject/pkg/apis/some.group.com/v1alpha1` => `some.group.comv1alpha1`

Reply:
@joelanford fair point, there is precedence for not using a version dir:

https://godoc.org/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions#CustomResourceDefinition

We could have two versions of the string argument:

1. `--custom-api-import-path="github.com/myproject/pkg/apis/some.group.com/v1alpha1=somev1alpha1", in which we do not infer the import identifier and instead use everything after `=`.
2. `--custom-api-import-path="github.com/myproject/pkg/apis/some/v1alpha1", in which we combine the last two path elements into the identifier.

We can return an error if the import alias already exists.

Reply:
@LiliC @estroz Lastly, thoughts on the flag name?

- `--custom-api-import`
- `--custom-api-import-path`
- `--k8s-api-import`
- `--k8s-api-import-path`

I think I prefer either of the `--k8s*` names. I would think this would be used most often to create a controller for k8s native resources (and less often with custom import paths), and `--custom*` strikes me as more confusing for the native k8s case than `--k8s*` does for the custom case.

Does `-path` add much here? Or can we remove that suffix to shorten the flag name without losing meaning?

Thoughts?

Reply:
I'm fine with `--k8s-api-import` considering the use case being addressed, as long as we document what it can do. `-path` is unnecessary.

Reply:
I tried this out and realized I was misunderstanding things a little bit.

```
operator-sdk add controller --k8s-api-import --api-version=core/v1 --kind=Pod
```

This returned `Error: required flag(s) "api-version" not set` because it used `--api-version=core/v1` as the value of the `--k8s-api-import` path.

I was thinking we'd make it so that setting the flag explicitly but without a value would mean "import this GVK from the `k8s.io/api` path", and setting it with a value would use the value as a custom path, basically what you described in [this comment](https://github.com/operator-framework/operator-sdk/pull/1344#discussion_r278368687), except that we'd use the value as the full path and not a prefix.

It doesn't look like that's possible with `pflag.StringVar()`, so I think we'd have to implement a  custom flag struct that implements the `pflag.Value` interface.

WDYT?

Reply:
Ah I thought we'd scrapped that part. I'd rather be explicit about import path so the experience is consistent for all external API's.

Reply:
Yeah, that is simpler so I'm good with that. In that case, I think your original `--custom-api-import` flag name is best. Sorry for the confusion!

Reply:
Np!